meerkat/block/manager.py

```diff
@@ -0,0 +1,176 @@
+        self._columns: Dict[str, AbstractColumn] = {}
+        self._column_to_block_id: Dict[str, int] = {}
+        self._block_refs: Dict[int, BlockRef] = {}
+
+    def update(self, block_ref: BlockRef):
+        """data (): a single blockable object, potentially contains multiple
+        columns."""
+        # can't have the same column living in multiple managers, so we view
+        self._columns.update(
+            {name: column.view() for name, column in block_ref.items()}
+        )
+
+        block_id = id(block_ref.block)
+        # check if there already is a block_ref in the manager for this block
+        if block_id in self._block_refs:
+            self._block_refs[block_id].update(block_ref)
+        else:
+            self._block_refs[block_id] = block_ref
+
+        self._column_to_block_id.update({name: block_id for name in block_ref.keys()})
+
+    def apply(self, method_name: str = "_get", *args, **kwargs) -> BlockManager:
+        """[summary]
+
+        Args:
+            fn (str): a function that is applied to a block and column_spec and
+                returns a new block and column_spec.
+        Returns:
+            [type]: [description]
+        """
+        results = None
+        for block_ref in self._block_refs.values():
+            result = block_ref.apply(method_name=method_name, *args, **kwargs)
+            if results is None:
+                results = BlockManager() if isinstance(result, BlockRef) else {}
+            results.update(result)
+
+        # apply method to columns not stored in block
+        for name, col in self._columns.items():
+            if results is not None and name in results:
+                continue
+
+            result = getattr(col, method_name)(*args, **kwargs)
+
+            if results is None:
+                results = BlockManager() if isinstance(result, AbstractColumn) else {}
+
+            results[name] = result
+
+        return results
+
+    def consolidate(self):
+        block_ref_groups = defaultdict(list)
+        for block_ref in self._block_refs.values():
+            block_ref_groups[block_ref.block.signature].append(block_ref)
+
+        for block_refs in block_ref_groups.values():
+            if len(block_refs) == 0:
+                continue
+
+            # remove old block_refs
+            for old_ref in block_refs:
+                del self._block_refs[id(old_ref.block)]
+
+            # consolidate group
+            block_class = block_refs[0].block.__class__
+            block_ref = block_class.consolidate(block_refs)
+            self.update(block_ref)
+
+    def remove(self, name):
+        if name not in self._columns:
+            raise ValueError(f"Remove failed: no column '{name}' in BlockManager.")
+
+        del self._columns[name]
+
+        if name in self._column_to_block_id:
+            # column is blockable
+            block_ref = self._block_refs[self._column_to_block_id[name]]
+            del block_ref[name]
```

Comment:
make this a pop (and everywhere else we do the del) – avoid garbage collecting directly